File: frontend/src/api/services/strategy.ts
/**
 * 策略开发 API 服务
 */
import { apiClient } from '../client';

// ============ 类型定义 ============

export interface Conversation {
  id: string;
  user_id: string;
  title: string;
  description?: string;
  status: 'active' | 'completed' | 'archived';
  current_strategy_id?: string;
  tags: string[];
  message_count: number;
  version_count: number;
  created_at: string;
  updated_at: string;
  last_message_at?: string;
}

export interface Message {
  id: string;
  conversation_id: string;
  role: 'user' | 'assistant' | 'system';
  content: string;
  generated_code?: string;
  code_language?: string;
  strategy_version_id?: string;
  metadata: Record<string, any>;
  created_at: string;
}

export interface StrategyVersion {
  id: string;
  strategy_name: string;
  user_id: string;
  conversation_id?: string;
  message_id?: string;
  code: string;
  code_encrypted: boolean;
  version: number;
  version_description?: string;
  strategy_type?: string;
  parameters: Record<string, any>;
  backtest_result_id?: string;
  performance_metrics?: Record<string, any>;
  status: 'draft' | 'testing' | 'active' | 'archived';
  created_at: string;
}

export interface StrategyTemplate {
  id: string;
  name: string;
  description: string;
  code: string;
  strategy_type: string;
  difficulty: 'beginner' | 'intermediate' | 'advanced';
  risk_level: 'low' | 'medium' | 'high';
  suitable_markets: string[];
  tags: string[];
  parameters: Record<string, any>;
  usage_count: number;
  created_by: string;
  is_public: boolean;
  created_at: string;
  updated_at: string;
}

// ============ 对话管理 API ============

export const conversationApi = {
  /**
   * 创建策略对话
   */
  create: async (data: {
    title: string;
    description?: string;
    tags?: string[];
  }): Promise<{ conversation_id: string; message: string }> => {
    const response = await apiClient.post('/strategy/conversations', data);
    return response.data;
  },

  /**
   * 获取对话列表
   */
  list: async (params?: {
    status?: 'active' | 'completed' | 'archived';
    skip?: number;
    limit?: number;
  }): Promise<{ conversations: Conversation[]; count: number }> => {
    const response = await apiClient.get('/strategy/conversations', { params });
    return response.data;
  },

  /**
   * 获取对话详情
   */
  get: async (conversationId: string): Promise<Conversation> => {
    const response = await apiClient.get(`/strategy/conversations/${conversationId}`);
    return response.data;
  },

  /**
   * 更新对话
   */
  update: async (
    conversationId: string,
    data: {
      title?: string;
      description?: string;
      tags?: string[];
      status?: 'active' | 'completed' | 'archived';
    }
  ): Promise<{ message: string }> => {
    const response = await apiClient.put(`/strategy/conversations/${conversationId}`, data);
    return response.data;
  },

  /**
   * 删除对话
   */
  delete: async (conversationId: string): Promise<{ message: string }> => {
    const response = await apiClient.delete(`/strategy/conversations/${conversationId}`);
    return response.data;
  },
};

// ============ 消息管理 API ============

export const messageApi = {
  /**
   * 添加消息
   */
  add: async (
    conversationId: string,
    data: {
      content: string;
      generated_code?: string;
      strategy_version_id?: string;
    }
  ): Promise<{ message_id: string; message: string }> => {
    const response = await apiClient.post(`/strategy/conversations/${conversationId}/messages`, data);
    return response.data;
  },

  /**
   * 获取消息列表
   */
  list: async (
    conversationId: string,
    params?: {
      skip?: number;
      limit?: number;
    }
  ): Promise<{ messages: Message[]; count: number }> => {
    const response = await apiClient.get(`/strategy/conversations/${conversationId}/messages`, {
      params,
    });
    return response.data;
  },
};

// ============ 策略版本管理 API ============

export const strategyApi = {
  /**
   * 创建策略版本
   */
  create: async (data: {
    strategy_name: string;
    code: string;
    conversation_id?: string;
    message_id?: string;
    strategy_type?: string;
    parameters?: Record<string, any>;
    version_description?: string;
  }): Promise<{ strategy_id: string; message: string }> => {
    const response = await apiClient.post('/strategy/strategies', data);
    return response.data;
  },

  /**
   * 获取策略版本列表
   */
  list: async (params?: {
    strategy_name?: string;
    skip?: number;
    limit?: number;
  }): Promise<{ strategies: StrategyVersion[]; count: number }> => {
    const response = await apiClient.get('/strategy/strategies', { params });
    return response.data;
  },

  /**
   * 获取策略版本详情
   */
  get: async (strategyId: string): Promise<StrategyVersion> => {
    const response = await apiClient.get(`/strategy/strategies/${strategyId}`);
    return response.data;
  },

  /**
   * 比较策略版本
   */
  compare: async (
    strategyId: string,
    targetId: string
  ): Promise<{
    version_1: any;
    version_2: any;
    code_diff: any;
  }> => {
    const response = await apiClient.get(`/strategy/strategies/${strategyId}/compare/${targetId}`);
    return response.data;
  },
};

// ============ 策略模板 API ============

export const templateApi = {
  /**
   * 获取模板列表
   */
  list: async (params?: {
    strategy_type?: string;
    difficulty?: 'beginner' | 'intermediate' | 'advanced';
    tags?: string[];
    skip?: number;
    limit?: number;
  }): Promise<{ templates: StrategyTemplate[]; count: number }> => {
    const response = await apiClient.get('/strategy/templates', { params });
    return response.data;
  },

  /**
   * 获取模板详情
   */
  get: async (templateId: string): Promise<StrategyTemplate> => {
    const response = await apiClient.get(`/strategy/templates/${templateId}`);
    return response.data;
  },

  /**
   * 搜索模板
   */
  search: async (keyword: string): Promise<{ templates: StrategyTemplate[]; count: number }> => {
    const response = await apiClient.get(`/strategy/templates/search/${keyword}`);
    return response.data;
  },
};

// ============ 回测管理 API ============

export interface BacktestResult {
  strategy_id: string;
  strategy_name: string;
  symbol: string;
  start_date: string;
  end_date: string;
  initial_cash: number;
  backtest_params: Record<string, any>;
  performance: {
    total_return: number;
    annual_return: number;
    sharpe_ratio: number;
    max_drawdown: number;
    win_rate: number;
    total_trades: number;
  };
  equity_curve: Array<{ date: string; value: number }>;
  trading_records: Array<any>;
  total_trades_count: number;
  created_at: string;
}

export const backtestApi = {
  /**
   * 运行策略回测
   */
  run: async (data: {
    strategy_id: string;
    symbol?: string;
    start_date?: string;
    end_date?: string;
    initial_cash?: number;
    backtest_params?: Record<string, any>;
  }): Promise<BacktestResult> => {
    const response = await apiClient.post('/strategy/backtest/run', data);
    return response.data;
  },

  /**
   * 快速回测（不保存策略版本）
   */
  quickRun: async (data: {
    strategy_code: string;
    symbol?: string;
    start_date?: string;
    end_date?: string;
    initial_cash?: number;
    strategy_params?: Record<string, any>;
  }): Promise<{
    performance: Record<string, any>;
    equity_curve: Array<any>;
    trading_records: Array<any>;
  }> => {
    const response = await apiClient.post('/strategy/backtest/quick', data);
    return response.data;
  },

  /**
   * 获取回测历史
   */
  getHistory: async (strategyId: string): Promise<{ backtest_results: BacktestResult[]; count: number }> => {
    const response = await apiClient.get(`/strategy/backtest/history/${strategyId}`);
    return response.data;
  },
};

File: backend/app/models/strategy.py
"""
策略开发 MongoDB 模型
"""
from datetime import datetime
from typing import Optional, Dict, Any, List
from pydantic import BaseModel, Field
from enum import Enum


class MessageRole(str, Enum):
    """消息角色"""
    USER = "user"
    ASSISTANT = "assistant"
    SYSTEM = "system"


class ConversationStatus(str, Enum):
    """对话状态"""
    ACTIVE = "active"  # 进行中
    COMPLETED = "completed"  # 已完成
    ARCHIVED = "archived"  # 已归档


class StrategyStatus(str, Enum):
    """策略状态"""
    DRAFT = "draft"  # 草稿
    TESTING = "testing"  # 测试中
    ACTIVE = "active"  # 激活
    ARCHIVED = "archived"  # 归档


class ConversationMessageModel(BaseModel):
    """对话消息MongoDB文档"""
    conversation_id: str = Field(..., description="对话ID")
    role: MessageRole = Field(..., description="消息角色")
    content: str = Field(..., description="消息内容")

    # 如果消息包含生成的代码
    generated_code: Optional[str] = Field(None, description="生成的策略代码")
    code_language: Optional[str] = Field("python", description="代码语言")

    # 关联的策略版本（如果此消息创建了新版本）
    strategy_version_id: Optional[str] = Field(None, description="关联的策略版本ID")

    # 元数据
    metadata: Dict[str, Any] = Field(default_factory=dict, description="消息元数据")

    # 时间戳
    created_at: datetime = Field(default_factory=datetime.now, description="创建时间")

    class Config:
        json_encoders = {
            datetime: lambda v: v.isoformat()
        }


class StrategyConversationModel(BaseModel):
    """策略对话MongoDB文档"""
    user_id: str = Field(..., description="用户ID")
    title: str = Field(..., description="对话标题")
    description: Optional[str] = Field(None, description="对话描述")

    # 对话状态
    status: ConversationStatus = Field(default=ConversationStatus.ACTIVE, description="对话状态")

    # 关联的策略（最新版本）
    current_strategy_id: Optional[str] = Field(None, description="当前策略ID")

    # 对话标签
    tags: List[str] = Field(default_factory=list, description="对话标签")

    # 统计信息
    message_count: int = Field(default=0, description="消息数量")
    version_count: int = Field(default=0, description="策略版本数量")

    # 时间戳
    created_at: datetime = Field(default_factory=datetime.now, description="创建时间")
    updated_at: datetime = Field(default_factory=datetime.now, description="更新时间")
    last_message_at: Optional[datetime] = Field(None, description="最后消息时间")

    class Config:
        json_encoders = {
            datetime: lambda v: v.isoformat()
        }


class StrategyVersionModel(BaseModel):
    """策略版本MongoDB文档"""
    strategy_name: str = Field(..., description="策略名称")
    user_id: str = Field(..., description="用户ID")

    # 关联对话
    conversation_id: Optional[str] = Field(None, description="关联的对话ID")
    message_id: Optional[str] = Field(None, description="生成此版本的消息ID")

    # 策略代码（加密存储）
    code: str = Field(..., description="策略代码")
    code_encrypted: bool = Field(default=False, description="代码是否加密")

    # 版本信息
    version: int = Field(default=1, description="版本号")
    version_description: Optional[str] = Field(None, description="版本说明")

    # 策略元数据
    strategy_type: Optional[str] = Field(None, description="策略类型（多头、CTA、网格等）")
    parameters: Dict[str, Any] = Field(default_factory=dict, description="策略参数")

    # 回测信息（如果已回测）
    backtest_result_id: Optional[str] = Field(None, description="关联的回测结果ID")
    performance_metrics: Optional[Dict[str, Any]] = Field(None, description="绩效指标快照")

    # 状态
    status: StrategyStatus = Field(default=StrategyStatus.DRAFT, description="策略状态")

    # 时间戳
    created_at: datetime = Field(default_factory=datetime.now, description="创建时间")

    class Config:
        json_encoders = {
            datetime: lambda v: v.isoformat()
        }


class StrategyTemplateModel(BaseModel):
    """策略模板MongoDB文档"""
    name: str = Field(..., description="模板名称")
    description: str = Field(..., description="模板描述")

    # 模板代码
    code: str = Field(..., description="模板代码")

    # 模板元数据
    strategy_type: str = Field(..., description="策略类型")
    difficulty: str = Field(default="intermediate", description="难度级别: beginner/intermediate/advanced")
    risk_level: str = Field(default="medium", description="风险等级: low/medium/high")

    # 适用场景
    suitable_markets: List[str] = Field(default_factory=list, description="适用市场")
    tags: List[str] = Field(default_factory=list, description="标签")

    # 参数说明
    parameters: Dict[str, Any] = Field(default_factory=dict, description="参数说明")

    # 使用统计
    usage_count: int = Field(default=0, description="使用次数")

    # 创建者
    created_by: str = Field(default="system", description="创建者（system或用户ID）")
    is_public: bool = Field(default=True, description="是否公开")

    # 时间戳
    created_at: datetime = Field(default_factory=datetime.now, description="创建时间")
    updated_at: datetime = Field(default_factory=datetime.now, description="更新时间")

    class Config:
        json_encoders = {
            datetime: lambda v: v.isoformat()
        }
